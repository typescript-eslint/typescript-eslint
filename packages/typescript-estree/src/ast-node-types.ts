/**
 * @fileoverview The AST node types produced by the parser.
 * @author Nicholas C. Zakas
 * @author James Henry <https://github.com/JamesHenry>
 * @copyright jQuery Foundation and other contributors, https://jquery.org/
 * MIT License
 */
export const AST_NODE_TYPES: { [key: string]: string } = {
  ArrayExpression: 'ArrayExpression',
  ArrayPattern: 'ArrayPattern',
  ArrowFunctionExpression: 'ArrowFunctionExpression',
  AssignmentExpression: 'AssignmentExpression',
  AssignmentPattern: 'AssignmentPattern',
  AwaitExpression: 'AwaitExpression',
  BigIntLiteral: 'BigIntLiteral',
  BinaryExpression: 'BinaryExpression',
  BlockStatement: 'BlockStatement',
  BreakStatement: 'BreakStatement',
  CallExpression: 'CallExpression',
  CatchClause: 'CatchClause',
  ClassBody: 'ClassBody',
  ClassDeclaration: 'ClassDeclaration',
  ClassExpression: 'ClassExpression',
  ClassImplements: 'ClassImplements',
  ClassProperty: 'ClassProperty',
  ConditionalExpression: 'ConditionalExpression',
  ContinueStatement: 'ContinueStatement',
  DebuggerStatement: 'DebuggerStatement',
  Decorator: 'Decorator',
  DoWhileStatement: 'DoWhileStatement',
  EmptyStatement: 'EmptyStatement',
  ExportAllDeclaration: 'ExportAllDeclaration',
  ExportDefaultDeclaration: 'ExportDefaultDeclaration',
  ExportNamedDeclaration: 'ExportNamedDeclaration',
  ExportSpecifier: 'ExportSpecifier',
  ExpressionStatement: 'ExpressionStatement',
  ForInStatement: 'ForInStatement',
  ForOfStatement: 'ForOfStatement',
  ForStatement: 'ForStatement',
  FunctionDeclaration: 'FunctionDeclaration',
  FunctionExpression: 'FunctionExpression',
  GenericTypeAnnotation: 'GenericTypeAnnotation',
  Identifier: 'Identifier',
  IfStatement: 'IfStatement',
  Import: 'Import',
  ImportDeclaration: 'ImportDeclaration',
  ImportDefaultSpecifier: 'ImportDefaultSpecifier',
  ImportNamespaceSpecifier: 'ImportNamespaceSpecifier',
  ImportSpecifier: 'ImportSpecifier',
  JSXAttribute: 'JSXAttribute',
  JSXClosingElement: 'JSXClosingElement',
  JSXClosingFragment: 'JSXClosingFragment',
  JSXElement: 'JSXElement',
  JSXEmptyExpression: 'JSXEmptyExpression',
  JSXExpressionContainer: 'JSXExpressionContainer',
  JSXFragment: 'JSXFragment',
  JSXIdentifier: 'JSXIdentifier',
  JSXMemberExpression: 'JSXMemberExpression',
  JSXNamespacedName: 'JSXNamespacedName',
  JSXOpeningElement: 'JSXOpeningElement',
  JSXOpeningFragment: 'JSXOpeningFragment',
  JSXSpreadAttribute: 'JSXSpreadAttribute',
  JSXSpreadChild: 'JSXSpreadChild',
  JSXText: 'JSXText',
  LabeledStatement: 'LabeledStatement',
  Literal: 'Literal',
  LogicalExpression: 'LogicalExpression',
  MemberExpression: 'MemberExpression',
  MetaProperty: 'MetaProperty',
  MethodDefinition: 'MethodDefinition',
  NewExpression: 'NewExpression',
  ObjectExpression: 'ObjectExpression',
  ObjectPattern: 'ObjectPattern',
  Program: 'Program',
  Property: 'Property',
  RestElement: 'RestElement',
  ReturnStatement: 'ReturnStatement',
  SequenceExpression: 'SequenceExpression',
  SpreadElement: 'SpreadElement',
  Super: 'Super',
  SwitchCase: 'SwitchCase',
  SwitchStatement: 'SwitchStatement',
  TaggedTemplateExpression: 'TaggedTemplateExpression',
  TemplateElement: 'TemplateElement',
  TemplateLiteral: 'TemplateLiteral',
  ThisExpression: 'ThisExpression',
  ThrowStatement: 'ThrowStatement',
  TryStatement: 'TryStatement',
  UnaryExpression: 'UnaryExpression',
  UpdateExpression: 'UpdateExpression',
  VariableDeclaration: 'VariableDeclaration',
  VariableDeclarator: 'VariableDeclarator',
  WhileStatement: 'WhileStatement',
  WithStatement: 'WithStatement',
  YieldExpression: 'YieldExpression',
  /**
   * TS-prefixed nodes
   */
  TSAbstractClassDeclaration: 'TSAbstractClassDeclaration',
  TSAbstractClassProperty: 'TSAbstractClassProperty',
  TSAbstractKeyword: 'TSAbstractKeyword',
  TSAbstractMethodDefinition: 'TSAbstractMethodDefinition',
  TSAnyKeyword: 'TSAnyKeyword',
  TSArrayType: 'TSArrayType',
  TSAsExpression: 'TSAsExpression',
  TSAsyncKeyword: 'TSAsyncKeyword',
  TSBooleanKeyword: 'TSBooleanKeyword',
  TSBigIntKeyword: 'TSBigIntKeyword',
  TSConditionalType: 'TSConditionalType',
  TSConstructorType: 'TSConstructorType',
  TSConstructSignature: 'TSConstructSignature',
  TSDeclareKeyword: 'TSDeclareKeyword',
  TSDeclareFunction: 'TSDeclareFunction',
  TSEnumDeclaration: 'TSEnumDeclaration',
  TSEnumMember: 'TSEnumMember',
  TSExportAssignment: 'TSExportAssignment',
  TSExportKeyword: 'TSExportKeyword',
  TSImportType: 'TSImportType',
  TSInferType: 'TSInferType',
  TSLiteralType: 'TSLiteralType',
  TSIndexedAccessType: 'TSIndexedAccessType',
  TSIndexSignature: 'TSIndexSignature',
  TSInterfaceBody: 'TSInterfaceBody',
  TSInterfaceDeclaration: 'TSInterfaceDeclaration',
  TSInterfaceHeritage: 'TSInterfaceHeritage',
  TSFunctionType: 'TSFunctionType',
  TSMethodSignature: 'TSMethodSignature',
  TSModuleBlock: 'TSModuleBlock',
  TSModuleDeclaration: 'TSModuleDeclaration',
  TSNamespaceFunctionDeclaration: 'TSNamespaceFunctionDeclaration',
  TSNonNullExpression: 'TSNonNullExpression',
  TSNeverKeyword: 'TSNeverKeyword',
  TSNullKeyword: 'TSNullKeyword',
  TSNumberKeyword: 'TSNumberKeyword',
  TSMappedType: 'TSMappedType',
  TSObjectKeyword: 'TSObjectKeyword',
  TSParameterProperty: 'TSParameterProperty',
  TSPrivateKeyword: 'TSPrivateKeyword',
  TSPropertySignature: 'TSPropertySignature',
  TSProtectedKeyword: 'TSProtectedKeyword',
  TSPublicKeyword: 'TSPublicKeyword',
  TSQualifiedName: 'TSQualifiedName',
  TSQuestionToken: 'TSQuestionToken',
  TSReadonlyKeyword: 'TSReadonlyKeyword',
  TSRestType: 'TSRestType',
  TSStaticKeyword: 'TSStaticKeyword',
  TSStringKeyword: 'TSStringKeyword',
  TSSymbolKeyword: 'TSSymbolKeyword',
  TSThisType: 'TSThisType',
  TSTypeAnnotation: 'TSTypeAnnotation',
  TSTypeAliasDeclaration: 'TSTypeAliasDeclaration',
  TSTypeLiteral: 'TSTypeLiteral',
  TSTypeOperator: 'TSTypeOperator',
  TSTypeParameter: 'TSTypeParameter',
  TSTypeParameterDeclaration: 'TSTypeParameterDeclaration',
  TSTypeParameterInstantiation: 'TSTypeParameterInstantiation',
  TSTypePredicate: 'TSTypePredicate',
  TSTypeReference: 'TSTypeReference',
  TSTypeQuery: 'TSTypeQuery',
  TSIntersectionType: 'TSIntersectionType',
  TSTupleType: 'TSTupleType',
  TSOptionalType: 'TSOptionalType',
  TSParenthesizedType: 'TSParenthesizedType',
  TSUnionType: 'TSUnionType',
  TSUndefinedKeyword: 'TSUndefinedKeyword',
  TSUnknownKeyword: 'TSUnknownKeyword',
  TSVoidKeyword: 'TSVoidKeyword'
};
