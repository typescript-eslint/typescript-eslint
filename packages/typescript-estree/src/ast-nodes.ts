import * as es from './ast-tree-nodes';

export type ASTTreeNodes =
  | es.ArrayExpression
  | es.ArrayPattern
  | es.ArrowFunctionExpression
  | es.AssignmentExpression
  | es.AssignmentPattern
  | es.AwaitExpression
  | es.BigIntLiteral
  | es.BinaryExpression
  | es.BlockStatement
  | es.BreakStatement
  | es.CallExpression
  | es.CatchClause
  | es.ClassBody
  | es.ClassDeclaration
  | es.ClassExpression
  | es.ClassProperty
  | es.ConditionalExpression
  | es.ContinueStatement
  | es.DebuggerStatement
  | es.Decorator
  | es.DoWhileStatement
  | es.EmptyStatement
  | es.ExportAllDeclaration
  | es.ExportDefaultDeclaration
  | es.ExportNamedDeclaration
  | es.ExportSpecifier
  | es.ExpressionStatement
  | es.ForInStatement
  | es.ForOfStatement
  | es.ForStatement
  | es.FunctionDeclaration
  | es.FunctionExpression
  | es.Identifier
  | es.IfStatement
  | es.Import
  | es.ImportDeclaration
  | es.ImportDefaultSpecifier
  | es.ImportNamespaceSpecifier
  | es.ImportSpecifier
  | es.JSXAttribute
  | es.JSXClosingElement
  | es.JSXClosingFragment
  | es.JSXElement
  | es.JSXEmptyExpression
  | es.JSXExpressionContainer
  | es.JSXFragment
  | es.JSXIdentifier
  | es.JSXMemberExpression
  // | es.JSXNamespacedName // https://github.com/Microsoft/TypeScript/issues/7411
  | es.JSXOpeningElement
  | es.JSXOpeningFragment
  | es.JSXSpreadAttribute
  | es.JSXSpreadChild
  | es.JSXText
  | es.LabeledStatement
  | es.Literal
  | es.LogicalExpression
  | es.MemberExpression
  | es.MetaProperty
  | es.MethodDefinition
  | es.NewExpression
  | es.ObjectExpression
  | es.ObjectPattern
  | es.Program
  | es.Property
  | es.RestElement
  | es.ReturnStatement
  | es.SequenceExpression
  | es.SpreadElement
  | es.Super
  | es.SwitchCase
  | es.SwitchStatement
  | es.TaggedTemplateExpression
  | es.TemplateElement
  | es.TemplateLiteral
  | es.ThisExpression
  | es.ThrowStatement
  | es.TryStatement
  | es.UnaryExpression
  | es.UpdateExpression
  | es.VariableDeclaration
  | es.VariableDeclarator
  | es.WhileStatement
  | es.WithStatement
  | es.YieldExpression
  | es.TSAbstractClassProperty
  // | es.TSAbstractKeyword
  | es.TSAbstractMethodDefinition
  | es.TSAnyKeyword
  | es.TSArrayType
  | es.TSAsExpression
  | es.TSAsyncKeyword
  | es.TSBooleanKeyword
  | es.TSBigIntKeyword
  | es.TSConditionalType
  | es.TSConstructorType
  | es.TSCallSignatureDeclaration
  | es.TSClassImplements
  | es.TSConstructSignatureDeclaration
  // | es.TSDeclareKeyword
  | es.TSDeclareFunction
  | es.TSEnumDeclaration
  | es.TSEnumMember
  | es.TSExportAssignment
  // | es.TSExportKeyword
  | es.TSExternalModuleReference
  | es.TSImportType
  | es.TSInferType
  | es.TSLiteralType
  | es.TSIndexedAccessType
  | es.TSIndexSignature
  | es.TSInterfaceBody
  | es.TSInterfaceDeclaration
  | es.TSInterfaceHeritage
  | es.TSImportEqualsDeclaration
  | es.TSFunctionType
  | es.TSMethodSignature
  | es.TSModuleBlock
  | es.TSModuleDeclaration
  | es.TSNamespaceExportDeclaration
  | es.TSNonNullExpression
  | es.TSNeverKeyword
  | es.TSNullKeyword
  | es.TSNumberKeyword
  | es.TSMappedType
  | es.TSObjectKeyword
  | es.TSParameterProperty
  | es.TSPrivateKeyword
  | es.TSPropertySignature
  | es.TSProtectedKeyword
  | es.TSPublicKeyword
  | es.TSQualifiedName
  // | es.TSQuestionToken
  // | es.TSReadonlyKeyword
  | es.TSRestType
  | es.TSStaticKeyword
  | es.TSStringKeyword
  | es.TSSymbolKeyword
  | es.TSThisType
  | es.TSTypeAnnotation
  | es.TSTypeAliasDeclaration
  | es.TSTypeAssertion
  | es.TSTypeLiteral
  | es.TSTypeOperator
  | es.TSTypeParameter
  | es.TSTypeParameterDeclaration
  | es.TSTypeParameterInstantiation
  | es.TSTypePredicate
  | es.TSTypeReference
  | es.TSTypeQuery
  | es.TSIntersectionType
  | es.TSTupleType
  | es.TSOptionalType
  | es.TSParenthesizedType
  | es.TSUnionType
  | es.TSUndefinedKeyword
  | es.TSUnknownKeyword
  | es.TSVoidKeyword
  // abstract types
  | es.FunctionDeclarationLike
  | es.ClassPropertyLikeNode
  | es.ClassLikeNode
  | es.TSFunctionDeclarationLike
  | es.TSExpressionWithTypeArgumentsLike;
