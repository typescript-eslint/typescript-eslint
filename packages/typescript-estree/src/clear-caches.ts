import { clearWatchCaches } from './create-program/getWatchProgramsForProjects';
import { clearProgramCache as clearProgramCacheOriginal } from './parser';
import { clearTSConfigMatchCache } from './parseSettings/createParseSettings';
import { clearGlobCache } from './parseSettings/resolveProjectList';

/**
 * Clears all of the internal caches.
 * Generally you shouldn't need or want to use this.
 * Examples of intended uses:
 * - In tests to reset parser state to keep tests isolated.
 * - In custom lint tooling that iteratively lints one project at a time to prevent OOMs.
 */
export function clearCaches(): void {
  clearProgramCacheOriginal();
  clearWatchCaches();
  clearTSConfigMatchCache();
  clearGlobCache();
}

// TODO - delete this in next major
export const clearProgramCache = clearCaches();
