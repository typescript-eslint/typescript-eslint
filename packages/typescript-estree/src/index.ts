export * from './parser';
export { ParserServices, TSESTreeOptions } from './parser-options';
export { simpleTraverse } from './simple-traverse';
export * from './ts-estree';
export { clearCaches } from './create-program/createWatchProgram';

// re-export for backwards-compat
export { visitorKeys } from '@typescript-eslint/visitor-keys';

// note - cannot migrate this to an import statement because it will make TSC copy the package.json to the dist folder
export const version: string = require('../package.json').version;
