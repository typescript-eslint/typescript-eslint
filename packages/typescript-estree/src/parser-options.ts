import { Program } from 'typescript';
import { Token, Comment, Node } from './ts-estree/ts-estree';
import { TSNode } from './ts-nodes';

export interface Extra {
  errorOnUnknownASTType: boolean;
  errorOnTypeScriptSyntacticAndSemanticIssues: boolean;
  useJSXTextNode: boolean;
  tokens: null | Token[];
  comment: boolean;
  code: string;
  range: boolean;
  loc: boolean;
  comments: Comment[];
  strict: boolean;
  jsx: boolean;
  log: Function;
  projects: string[];
  tsconfigRootDir: string;
  extraFileExtensions: string[];
}

export interface ParserOptions {
  range?: boolean;
  loc?: boolean;
  tokens?: boolean;
  comment?: boolean;
  jsx?: boolean;
  errorOnUnknownASTType?: boolean;
  errorOnTypeScriptSyntacticAndSemanticIssues?: boolean;
  useJSXTextNode?: boolean;
  loggerFn?: Function | false;
  project?: string | string[];
  filePath?: string;
  tsconfigRootDir?: string;
  extraFileExtensions?: string[];
}

export interface ParserWeakMap<TKey, TValueBase> {
  get<TValue extends TValueBase>(key: TKey): TValue;
  has(key: any): boolean;
}

export interface ParserServices {
  program: Program | undefined;
  esTreeNodeToTSNodeMap: ParserWeakMap<Node, TSNode> | undefined;
  tsNodeToESTreeNodeMap: ParserWeakMap<TSNode, Node> | undefined;
}
