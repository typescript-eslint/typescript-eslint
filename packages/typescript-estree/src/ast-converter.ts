import { SourceFile } from 'typescript';
import { convertError, Converter, ASTMaps } from './convert';
import { convertComments } from './convert-comments';
import { convertTokens } from './node-utils';
import { Extra } from './parser-options';
import { TSESTree } from './ts-estree';
import { simpleTraverse } from './simple-traverse';

export function astConverter(
  ast: SourceFile,
  extra: Extra,
  shouldPreserveNodeMaps: boolean,
): { estree: TSESTree.Program; astMaps: ASTMaps | undefined } {
  /**
   * The TypeScript compiler produced fundamental parse errors when parsing the
   * source.
   */
  // internal typescript api...
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const parseDiagnostics = (ast as any).parseDiagnostics;
  if (parseDiagnostics.length) {
    throw convertError(parseDiagnostics[0]);
  }

  /**
   * Recursively convert the TypeScript AST into an ESTree-compatible AST
   */
  const instance = new Converter(ast, {
    errorOnUnknownASTType: extra.errorOnUnknownASTType || false,
    useJSXTextNode: extra.useJSXTextNode || false,
    shouldPreserveNodeMaps,
  });

  const estree = instance.convertProgram();

  /**
   * Optionally remove range and loc if specified
   */
  if (!extra.range || !extra.loc) {
    simpleTraverse(estree, {
      enter: node => {
        if (!extra.range) {
          delete node.range;
        }
        if (!extra.loc) {
          delete node.loc;
        }
      },
    });
  }

  /**
   * Optionally convert and include all tokens in the AST
   */
  if (extra.tokens) {
    estree.tokens = convertTokens(ast);
  }

  /**
   * Optionally convert and include all comments in the AST
   */
  if (extra.comment) {
    estree.comments = convertComments(ast, extra.code);
  }

  const astMaps = shouldPreserveNodeMaps ? instance.getASTMaps() : undefined;

  return { estree, astMaps };
}
