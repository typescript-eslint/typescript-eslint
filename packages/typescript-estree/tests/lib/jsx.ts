import { readFileSync } from 'fs';
import glob from 'glob';
import { ParserOptions } from '../../src/parser-options';
import {
  createSnapshotTestBlock,
  formatSnapshotName,
} from '../../tools/test-utils';
import filesWithKnownIssues from '../../../shared-fixtures/jsx-known-issues';

const JSX_FIXTURES_DIR =
  '../../node_modules/@typescript-eslint/shared-fixtures/fixtures/jsx';
const jsxTestFiles = glob
  .sync(`${JSX_FIXTURES_DIR}/**/*.src.js`)
  .filter(filename =>
    filesWithKnownIssues.every(fileName => !filename.includes(fileName)),
  );

const JSX_JSXTEXT_FIXTURES_DIR =
  '../../node_modules/@typescript-eslint/shared-fixtures/fixtures/jsx-useJSXTextNode';
const jsxTextTestFiles = glob.sync(`${JSX_JSXTEXT_FIXTURES_DIR}/**/*.src.js`);

describe('JSX', () => {
  /**
   * Test each fixture file
   */
  function testFixture(
    fixturesDir: string,
    useJSXTextNode: boolean,
  ): (filename: string) => void {
    return filename => {
      const code = readFileSync(filename, 'utf8');
      const config: ParserOptions = {
        loc: true,
        range: true,
        tokens: true,
        errorOnUnknownASTType: true,
        useJSXTextNode,
        jsx: true,
      };
      it(
        formatSnapshotName(filename, fixturesDir),
        createSnapshotTestBlock(code, config),
      );
    };
  }

  describe('useJSXTextNode: false', () => {
    jsxTestFiles.forEach(testFixture(JSX_FIXTURES_DIR, false));
  });
  describe('useJSXTextNode: true', () => {
    jsxTextTestFiles.forEach(testFixture(JSX_JSXTEXT_FIXTURES_DIR, true));
  });
});
