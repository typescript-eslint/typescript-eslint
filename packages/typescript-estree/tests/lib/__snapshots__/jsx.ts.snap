// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSX useJSXTextNode: false fixtures/invalid-attribute.src 1`] = `"'{' expected."`;

exports[`JSX useJSXTextNode: false fixtures/invalid-attribute-missing-equals.src 1`] = `"Identifier expected."`;

exports[`JSX useJSXTextNode: false fixtures/invalid-broken-tag.src 1`] = `"Unterminated string literal."`;

exports[`JSX useJSXTextNode: false fixtures/invalid-computed-end-tag-name.src 1`] = `"Identifier expected."`;

exports[`JSX useJSXTextNode: false fixtures/invalid-computed-string-end-tag-name.src 1`] = `"Identifier expected."`;

exports[`JSX useJSXTextNode: false fixtures/invalid-embedded-expression.src 1`] = `"'}' expected."`;

exports[`JSX useJSXTextNode: false fixtures/invalid-leading-dot-tag-name.src 1`] = `"Expression expected."`;

exports[`JSX useJSXTextNode: false fixtures/invalid-matching-placeholder-in-closing-tag.src 1`] = `"'>' expected."`;

exports[`JSX useJSXTextNode: false fixtures/invalid-mismatched-closing-tag.src 1`] = `"Expected corresponding JSX closing tag for 'a'."`;

exports[`JSX useJSXTextNode: false fixtures/invalid-mismatched-closing-tags.src 1`] = `"JSX element 'a' has no corresponding closing tag."`;

exports[`JSX useJSXTextNode: false fixtures/invalid-mismatched-dot-tag-name.src 1`] = `"Expected corresponding JSX closing tag for 'a.b.c'."`;

exports[`JSX useJSXTextNode: false fixtures/invalid-mismatched-namespace-tag.src 1`] = `"Identifier expected."`;

exports[`JSX useJSXTextNode: false fixtures/invalid-missing-closing-tag.src 1`] = `"JSX element 'a' has no corresponding closing tag."`;

exports[`JSX useJSXTextNode: false fixtures/invalid-missing-closing-tag-attribute-placeholder.src 1`] = `"JSX element 'a' has no corresponding closing tag."`;

exports[`JSX useJSXTextNode: false fixtures/invalid-missing-namespace-name.src 1`] = `"Expression expected."`;

exports[`JSX useJSXTextNode: false fixtures/invalid-missing-namespace-value.src 1`] = `"Identifier expected."`;

exports[`JSX useJSXTextNode: false fixtures/invalid-missing-spread-operator.src 1`] = `"'...' expected."`;

exports[`JSX useJSXTextNode: false fixtures/invalid-namespace-name-with-docts.src 1`] = `"Identifier expected."`;

exports[`JSX useJSXTextNode: false fixtures/invalid-no-common-parent.src 1`] = `"JSX expressions must have one parent element."`;

exports[`JSX useJSXTextNode: false fixtures/invalid-no-common-parent-with-comment.src 1`] = `"JSX expressions must have one parent element."`;

exports[`JSX useJSXTextNode: false fixtures/invalid-no-tag-name.src 1`] = `"Declaration or statement expected."`;

exports[`JSX useJSXTextNode: false fixtures/invalid-placeholder-in-closing-tag.src 1`] = `"'>' expected."`;

exports[`JSX useJSXTextNode: false fixtures/invalid-shorthand-fragment-no-closing.src 1`] = `"JSX fragment has no corresponding closing tag."`;

exports[`JSX useJSXTextNode: false fixtures/invalid-trailing-dot-tag-name.src 1`] = `"Identifier expected."`;

exports[`JSX useJSXTextNode: false fixtures/invalid-unexpected-comma.src 1`] = `"Identifier expected."`;

exports[`JSX useJSXTextNode: false fixtures/newslines-and-entities.src 1`] = `"Invalid character."`;

exports[`JSX useJSXTextNode: false fixtures/self-closing-tag-with-newline.src 1`] = `"Invalid character."`;
